### ISSUE TITLE
vinyl: rework upsert operation

### ISSUE THREAD

**Issue by Korablev77** [tarantool/tarantool#5107]
There are several issues and bugs connected with upsert operation in vinyl. Let's sum them up in one place.
To start with let's define upsert specification (i.e. how it is supposed to work):
`space_object:upsert({tuple}, {{operator, field_no, value}, ...})`
Depending on existence of primary_key(tuple) there are two main cases:
1. If there is an existing tuple which matches the key fields of tuple, then the request has the same effect as `space_object:update()` and the` {{operator, field_no, value}, ...}` parameter is used.
2. If there is no existing tuple which matches the key fields of tuple, then the request has the same effect as `space_object:insert()` and the `{tuple}` parameter is used.
There are also two processes under the hood: execute and squash. Both are handled in `vy_apply_upsert()`. Assume we have two upserts:
```
upsert1 = {tuple1, {ops1}} -- old one, i.e. lsn(upsert1) <= lsn(upsert2)
upsert2 = {tuple2, {ops2}} -- new one
```
Execute procedure applies ops2 on tuple1. Operations which can't be applied are skipped. Note that if upsert folds into update, then execute procedure is redundant (for details see https://github.com/tarantool/tarantool/issues/5092).
Squash procedure combines ops1 with ops2. For instance, if both ops contain arithmetic operations involving the same field - they are simplified to one operation. In case squash is not possible, both ops are written in a single array, so that after squash we have single upsert.
Instead of upsert1 upsert2 can be applied on any of DELETE/INSERT/REPLACE statements:
 - if statement is delete, upsert2 turns into replace storing tuple1;
 - if statement is insert/replace, upsert2 turns into update and ops2 are applied on the tuple stored in old statement.
After squash procedure resulting statement is verified to not change primary key: if resulting statement has different PK from upsert1, then procedure is rolled back and old statement is returned instead **(*)**.

What's wrong with upsert now:
 1. if upsert1 holds '=' operation and upsert2 holds any other arithmetic operation over the same field then the result of squash is broken (https://github.com/tarantool/tarantool/issues/5106);
 2. if upsert2 refers to -1 field (which means the last field in the tuple) then upsert may be skipped (https://github.com/tarantool/tarantool/issues/5087);
 3. if execute procedure fails (e.g. due to OOM; note that operation which can't be execute is not considered to be OK and is simply skipped) it may lead to crash (https://github.com/tarantool/tarantool/issues/4957);
 4. due to upserts squashing final result may be completely skipped even despite the fact that some upserts are valid (https://github.com/tarantool/tarantool/issues/5105). In fact, this problem is very similar to 2 point from this list: it appears since update operations may modify PK of tuple1 (even if upsert folds into update and tuple1 is not used at all) - see **(*)**;
5. tuple format check takes place only once during initial insertion of upsert statement into L0. Note that this check is redundant if upsert folds into update. Moreover, there's no verification that after upsert execution resulting tuple will match tuple format (https://github.com/tarantool/tarantool/issues/1622);
6. skip upserts which can't be applied. Now they get stuck in mem/runs until replace/delete statement is met preventing dump/compaction processes from being finished. Moreover, let's log such upserts for the sake of diagnostic simplicity.

Minor&performance issues:
 - defer upsert execution till we certainly know that upsert folds into insert (https://github.com/tarantool/tarantool/issues/5106). Note that we should mark update operations corresponding to first upsert in series since if upsert folds into insert we should not apply its update operation. Furthermore, this may require changing upsert storage format (this mark should be persisted).
 - if squash fails, resulting upsert will contain more update operations than upsert1 and upsert2. After squashing long series of upserts (squash takes place during explicit read or compaction), we may get upsert with number of update operations more than hardcoded threshold - `BOX_UPDATE_OP_CNT_MAX == 4000`. We have nothing to do but skip such upsert (ergo the whole series of upserts)  since we can't operate on such giant number of operations. I believe we should better attempt at seeking the terminal statement and apply such upsert on it and then squash the rest of upserts in series.

What should be done to resolve these problems:

- [x] Fix and test 1. point from the list. Problem seems to be relatively independent from other.
- [x] Fix and test 3. point from the list. The only thing to do here is add one NULL check (possible fix is here: https://github.com/tarantool/tarantool/commit/776e8b91b93c79dabd2932b5d665236c5da313c8).
- [ ] Deal with tuple format checks: remove one from `vy_upsert` and instead add it to `vy_apply_upsert` (after upsert is applied to the terminal statement or folds into insert).
- [ ] Skip&log upserts which can't be applied. PoC here: https://github.com/tarantool/tarantool/commit/52940407b621019b10a1fa79850388fbc2e17d4b
- [ ] Deal with 4 and 2: they are both about tuple PK modification. PoC is here: https://github.com/tarantool/tarantool/commit/7c98fabb09bef52546b592ce5603e7e69b3ffca1 The main drawback of suggested solution is that still whole upsert is skipped if one of operations _really_ modifies PK (in case upsert folds into update). Another possible solution is to firstly implement deferred upsert execution. In this case we would be able to skip only those update operations which modify PK (otherwise we can't rollback effects of upsert execution).
- [ ] Allow applying upserts with number of operations more than threshold (maybe as configurable option).

Suggestions, clarifications and comments are welcome.

**Comment by Korablev77** [comment]
Closed by https://github.com/tarantool/tarantool/commit/5a61c4717e63de24359a02c9f3e603444ab5b637
Upserts squash optimization follow-up: https://github.com/tarantool/tarantool/issues/5428
